main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -420,59 +429,79 @@ object Lexer {
       case _ if isKeyword("select") => TokenKind.KeywordSelect
+      case _ if isKeyword("set") => TokenKind.KeywordSet
       case _ if isKeyword("solve") => TokenKind.KeywordSolve
       case _ if isKeyword("spawn") => TokenKind.KeywordSpawn
       case _ if isKeyword("static") => TokenKind.KeywordStatic
+      case _ if isKeyword("Static") => TokenKind.KeywordStaticUppercase
+      case _ if isKeyword("trait") => TokenKind.KeywordTrait
       case _ if isKeyword("true") => TokenKind.KeywordTrue
       case _ if isKeyword("try") => TokenKind.KeywordTry
       case _ if isKeyword("type") => TokenKind.KeywordType
       case _ if isKeyword("typematch") => TokenKind.KeywordTypeMatch
       case _ if isKeyword("unchecked_cast") => TokenKind.KeywordUncheckedCast
-      case _ if isKeyword("univ") => TokenKind.KeywordUniv
+      case _ if isKeyword("Univ") => TokenKind.KeywordUniv
       case _ if isKeyword("use") => TokenKind.KeywordUse
       case _ if isKeyword("where") => TokenKind.KeywordWhere
       case _ if isKeyword("with") => TokenKind.KeywordWith
       case _ if isKeyword("without") => TokenKind.KeywordWithout
       case _ if isKeyword("yield") => TokenKind.KeywordYield
+      case _ if isKeyword("xor") => TokenKind.KeywordXor
       case _ if isKeyword("Set#") => TokenKind.SetHash
       case _ if isKeyword("Array#") => TokenKind.ArrayHash
       case _ if isKeyword("Map#") => TokenKind.MapHash
       case _ if isKeyword("List#") => TokenKind.ListHash
       case _ if isKeyword("Vector#") => TokenKind.VectorHash
-      case _ if isKeyword("regex\"") => acceptRegex()
+      case _ if isMatch("regex\"") => acceptRegex()
       case _ if isMathNameChar(c) => acceptMathName()
       case _ if isGreekNameChar(c) => acceptGreekName()
+      case c if c.isLetter => acceptName(c.isUpper)
+      case c if DIGITS.contains(c) => acceptNumber()
       // User defined operators.
       case _ if ValidUserOpTokens.contains(c) =>
         val p = peek()
         if (ValidUserOpTokens.contains(p)) {
           acceptUserDefinedOp()
-        } else if (c == '-' && p.isDigit) {
-          acceptNumber() // negative numbers.
         } else {
           ValidUserOpTokens.apply(c)
         }
-      case c if c.isLetter => acceptName(c.isUpper)
-      case c if c.isDigit => acceptNumber()
       case c => TokenKind.Err(LexerError.UnexpectedChar(c.toString, sourceLocationAtStart()))
     }
   }
 
+  /**
+   * Check that the potential keyword is sufficiently separated, taking care not to go out-of-bounds.
+   * A keyword is separated if it is surrounded by whitespace, parenthesis, brackets or curlies.
+   * Note that __comparison includes current__.
+   */
+  private def isSeparated(keyword: String)(implicit s: State): Boolean = {
+    // TODO: We can be smarter!
+    val VALID_SEPARATORS =  List('(', ')', '[', ']', '{', '}', ',', ';', '~', '+', '-')
+    def isSep(c: Char) = c.isWhitespace || VALID_SEPARATORS.contains(c)
```

Comment:
The call to contains is also likely to be slow.

Reply:
I would just check if `c == ...`

Reply:
I think past me was trying to say that maybe the whole `VALID_SEPARATORS` idea is not the right approach.
> `// TODO: We can be smarter!`
I think I will investigate that a little, otherwise I can go for the `c == ... || c == ...` style.

Reply:
We are still about 2x slower, so I want to try and squeeze a bit more!
I didn't copy the doc-comment into the snippet. The full thing is
```scala
/**
   * Check that the potential keyword is sufficiently separated, taking care not to go out-of-bounds.
   * A keyword is separated if it is surrounded by anything __but__ a character, digit or underscore.
   * Note that __comparison includes current__.
   */
  private def isSeparated(keyword: String)(implicit s: State): Boolean = {
    def isSep(c: Char) = !(c.isWhitespace || c.isLetter || c.isDigit || c == '_')
    s.src.data.lift(s.current.offset - 2).forall(isSep) && s.src.data.lift(s.current.offset + keyword.length - 1).forall(isSep)
  }
```

Reply:
Ok. Is that significant for the lexer?

Reply:
We are talking 200ms vs 400ms on my computer. That's about 4% vs. 2%

Reply:
Does it mean you want me to hold of merging?

Reply:
Hmm. Yeah thats a bit in the high-end.